blog/_posts/2020-11-24-ranluxpp.md

```diff
@@ -0,0 +1,146 @@
+---
+title: "Implementing and tuning RANLUX++ for ROOT"
+layout: archive
+author: Jonas Hahnfeld
+---
+
+ROOT includes different implementations of pseudorandom number generators (PRNGs):
+ * [`TRandom`][TRandom] is a Linear Congruential Generator (LCG) based on the BSD `rand` generator.
+ * [`TRandom1`][TRandom1] implements the RANLUX generator described by Fred James in 1990.
+ * [`TRandom3`][TRandom3] is based on the Mersenne Twister generator.
+   By default, this implementation is used for `gRandom`.
+ * `TRandomMixMax` is an instantiation of the parametric [`MixMaxEngine`][MixMaxEngine].
+
+[TRandom]: https://root.cern/doc/master/classTRandom.html
+[TRandom1]: https://root.cern/doc/master/classTRandom1.html
+[Trandom3]: https://root.cern/doc/master/classTRandom3.html
+[MixMaxEngine]: https://root.cern/doc/master/classROOT_1_1Math_1_1MixMaxEngine.html
+
+This post discusses the implementation of the recently added `TRandomRanluxpp`.
+
+## Why so many PRNGs?
+
+There exist many properties to evaluate PRNGs, with the easiest being its period:
+It captures how many random numbers can be generated before the sequence "wraps around" and repeats.
+For obvious reasons, the period of a PRNG cannot be longer than its state.
+This is one of the problems of the oldest LCGs with a state of 32 bits or less.
+
+However, there are many more statistical properties that are at least equally important.
+A widely used collection of empirical tests is [TestU01][TestU01] by L'Ecuyer and Simard.
+It has to be noted that the tests can only prove the existence of defects.
+That is, passing all tests does not mean that a PRNG is free of errors.
+
+[TestU01]: http://simul.iro.umontreal.ca/testu01/tu01.html
+
+As with many other design issues, PRNGs need to make a certain trade-off:
+Fast generators with a small state will usually fail many statistical tests.
+On the other hand, good generators are typically slower or use a much larger state.
+In fact, most generators will fail one test or another, including the widely used Mersenne Twister.
+Notable exceptions are RANLUX and MIXMAX that pass all currently known tests.
+However, they are rather slow (RANLUX) or require a large state (MIXMAX).
+
+## RANLUX++
+
+[RANLUX++][paper] is an LCG implementation derived from the ideas of RANLUX.
+It aims to provide better performance while maintaining the good properties of RANLUX.
+The generator uses a single 576 bit number as state.
+Additionally it needs to remember from which offset to take the next random bits.
```

Comment:
I don't understand this (because I'm not expert enough on PRNGs, for sure, but consider I'm the average ROOT blog reader)

Reply:
Hopefully clarified a bit

Reply:
:+1: